src/dkg/round3.rs

```diff
@@ -0,0 +1,342 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::borrow::Borrow;
+use std::collections::BTreeMap;
+
+use crate::checksum::ChecksumError;
+use crate::frost::keys::dkg::round1::Package as Round1Package;
+use crate::frost::keys::dkg::round2::Package as Round2Package;
+use crate::frost::keys::dkg::round2::SecretPackage as Round2SecretPackage;
+use crate::participant::Identity;
+use reddsa::frost::redjubjub::keys::dkg::part3;
+use reddsa::frost::redjubjub::keys::KeyPackage;
+use reddsa::frost::redjubjub::keys::PublicKeyPackage;
+use reddsa::frost::redjubjub::Identifier;
+
+use super::error::Error;
+use super::round1;
+use super::round2;
+
+pub fn round3<'a, P, Q>(
+    identity: &Identity,
+    round2_secret_package: &Round2SecretPackage,
+    round1_public_packages: P,
+    round2_public_packages: Q,
+) -> Result<(KeyPackage, PublicKeyPackage), Error>
+where
+    P: IntoIterator<Item = &'a round1::PublicPackage>,
+    Q: IntoIterator<Item = &'a round2::PublicPackage>,
+{
+    let round1_public_packages = round1_public_packages.into_iter().collect::<Vec<_>>();
+
+    let (min_signers, max_signers) = round2::get_secret_package_signers(round2_secret_package);
+
+    // Ensure that the number of public packages provided matches max_signers
+    if round1_public_packages.len() != max_signers as usize {
+        return Err(Error::InvalidInput(format!(
+            "expected {} public packages, got {}",
+            max_signers,
+            round1_public_packages.len()
+        )));
+    }
+
+    let expected_round1_checksum = round1::input_checksum(
+        min_signers,
+        round1_public_packages.iter().map(|pkg| pkg.identity()),
+    );
+
+    let mut identities = BTreeMap::new();
```

Comment:
do we need the `identities` map for round3? it doesn't look like we read from it later on